tools/2024-06-gitops-feature-spec.md

```diff
@@ -0,0 +1,374 @@
+
+> As a Developer, I'm very happy with this experience because I get the cool new features of Radius (like self-serve infrastructure deployment via Recipes and the App Graph/Dashboard) in addition to the features I already know and love in Flux, like health monitoring and rollback.
+
+__Requirements resulting from these scenarios:__ 
+- Flux/ArgoCD must be able to detect the file change described above and must be able to read the env.bicep file as required to deploy the Radius app correctly 
+- Flux/ArgoCD must be able to detect changes to the `.bicepconfig` file and execute bicep config updates
+- Radius types and Bicep files must be able to be read by Flux/ArgoCD which currently does not have that ability.
+- Flux/ArgoCD must be able to detect Radius resource types and execute Flux/ArgoCD commands against those types, including rollback *list of any other required Flux commands*.
+- Based on user feedback this could be just testing that commands such as dry run mechanisms work
+
+## Key investments
+<!-- List the features required to enable this scenario. -->
+
+### Feature 1: GitOps tooling can detect and understand Radius definition files
+<!-- One or two sentence summary -->
+GitOps tools (beginning with Flux and ArgoCD) must be able to detect the file change in the git repo described above and must be able to read the `env.bicep` file (or any `*.bicep` and `.bicepconfig` files) as required to deploy the Radius app correctly.
+
+GitOps tools (beginning with Flux and ArgoCD) must be able to detect changes to Radius applications and other resource type definitions in the git repo and notify Radius of the changes.
+
+### Feature 2: GitOps tooling can trigger Radius to execute commands
+<!-- One or two sentence summary -->
+Once Radius definition file changes are detected and understood, GitOps tools (beginning with Flux and ArgoCD) must be able to notify and trigger Radius to execute deployments for the Radius resources defined in the git repo.
+
+### Feature 3: GitOps tooling can rollback Radius resources
+<!-- One or two sentence summary -->
+GitOps tools (beginning with Flux and ArgoCD) must be able to execute rollback operations for Radius resources defined in the git repo. In other words, the user may use the GitOps rollback commands or `git revert` to revert a change to a Radius resource and Flux/ArgoCD will detect the change and trigger Radius to redeploy the resource with the previous configuration.
+
+### Feature 4: GitOps tooling can detect and execute resource deletions
+<!-- One or two sentence summary -->
+If a resource in a `.bicep` Radius definition file is deleted and the change is pushed to the repo, the GitOps tools (Flux and ArgoCD) should detect the change and trigger Radius to delete that resource. Note that this is not a behavior of bicep today but it's important for this usecase.
```

Comment:
AFAIK we don't have an "undeploy" mechanism built in today. might take some thinking to enable this scenario

Reply:
"declarative deletion"

Reply:
added a note for this.